pkg/operator/resources/validations.go

```diff
@@ -134,83 +141,148 @@ var _inferentiaCPUReserve = kresource.MustParse("100m")
-			}
+	for _, ngName := range api.NodeGroups {
+		if !clusterNodeGroupNames.Has(ngName) {
+			return errors.Wrap(ErrorInvalidNodeGroupSelector(ngName, config.ClusterConfig.GetNodeGroupNames()), userconfig.NodeGroupsKey)
 		}
 	}
 
-	compute := userconfig.GetTotalComputeFromContainers(api.Pod.Containers)
+	compute := userconfig.GetPodComputeRequest(api)
 
-	for _, instanceMetadata := range config.InstancesMetadata {
-		if apiNodeGroupNames != nil {
-			matchedNodeGroups := 0
-			for _, ngName := range apiNodeGroupNames {
-				if config.ClusterConfig.GetNodeGroupByName(ngName).InstanceType == instanceMetadata.Type {
-					matchedNodeGroups++
-				}
-			}
-			if matchedNodeGroups == 0 {
-				continue
-			}
+	for _, ng := range config.ClusterConfig.NodeGroups {
+		if api.NodeGroups != nil && !slices.HasString(api.NodeGroups, ng.Name) {
+			continue
 		}
 
-		maxMemLoop := maxMemMap[instanceMetadata.Type]
-		maxMemLoop.Sub(_cortexMemReserve)
-
-		maxCPU := instanceMetadata.CPU
-		maxCPU.Sub(_cortexCPUReserve)
+		nodeCPU, nodeMem, nodeGPU, nodeInf := getNodeCapacity(ng.InstanceType, maxMemMap)
 
-		maxGPU := instanceMetadata.GPU
-		if maxGPU > 0 {
-			// Reserve resources for nvidia device plugin daemonset
-			maxCPU.Sub(_nvidiaCPUReserve)
-			maxMemLoop.Sub(_nvidiaMemReserve)
-			// Reserve resources for nvidia dcgm prometheus exporter
-			maxCPU.Sub(_nvidiaDCGMExporterCPUReserve)
-			maxMemLoop.Sub(_nvidiaDCGMExporterMemReserve)
+		if compute.CPU != nil && nodeCPU.Cmp(compute.CPU.Quantity) < 0 {
+			continue
+		} else if compute.Mem != nil && nodeMem.Cmp(compute.Mem.Quantity) < 0 {
+			continue
+		} else if compute.GPU > nodeGPU {
+			continue
+		} else if compute.Inf > nodeInf {
+			continue
 		}
 
-		maxInf := instanceMetadata.Inf
-		if maxInf > 0 {
-			// Reserve resources for inferentia device plugin daemonset
-			maxCPU.Sub(_inferentiaCPUReserve)
-			maxMemLoop.Sub(_inferentiaMemReserve)
-		}
+		// we found a node group that has capacity
+		return nil
+	}
 
-		loopErrors := []error{}
-		if compute.CPU != nil && maxCPU.Cmp(compute.CPU.Quantity) < 0 {
-			loopErrors = append(loopErrors, ErrorNoAvailableNodeComputeLimit("CPU", compute.CPU.String(), maxCPU.String()))
-		}
-		if compute.Mem != nil && maxMemLoop.Cmp(compute.Mem.Quantity) < 0 {
-			loopErrors = append(loopErrors, ErrorNoAvailableNodeComputeLimit("memory", compute.Mem.String(), maxMemLoop.String()))
-		}
-		if compute.GPU > maxGPU {
-			loopErrors = append(loopErrors, ErrorNoAvailableNodeComputeLimit("GPU", fmt.Sprintf("%d", compute.GPU), fmt.Sprintf("%d", maxGPU)))
+	// no nodegroups have capacity
+	errMsg := "no instance types in your cluster are large enough to satisfy the requested resources for your pod\n\n"
+	errMsg += console.Bold("requested pod resources\n")
+	errMsg += podResourceRequestsTable(api, compute)
+	errMsg += "\n" + s.TrimTrailingNewLines(nodeGroupResourcesTable(api, compute, maxMemMap))
+	return ErrorNoAvailableNodeComputeLimit(errMsg)
```

Comment:
Since `podResourceRequestsTable` and `nodeGroupResourcesTable` functions are only required for the `ErrorNoAvailableNodeComputeLimit` error, I think it might be cleaner to move these 2 functions inside the errors module and call them inside the `ErrorNoAvailableNodeComputeLimit` error function.

We'd return the error the following way then: `ErrorNoAvailableNodeComputeLimit(api, compute, maxMemMap)`.